File: HW/HW10/0318-HW10.py
from typing import List, Optional

class TreeNode:
    def __init__(self, value):
        self.value = value
        self.left = None
        self.right = None


    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        if root is None:
            return []
        
        result = []
        current = [root]

        while current:
            next_level = []
            values = []

            for node in current:
                values.append(node.val)
                if node.left: next_level.append(node.left)
                if node.right: next_level.append(node.right)

            result.append(values)
            current = next_level
        return result
    

root = TreeNode(4)
root.left = TreeNode(3)
root.left.left = TreeNode(1)
root.right = TreeNode(8)
root.right.left = TreeNode(5)
root.right.right = TreeNode(9)

print(root.levelOrder(root)) #testcase

File: HW/HW10/0318-HW10.java
import java.util.List;
import java.util.Queue;
import java.util.ArrayList;
import java.util.LinkedList;


class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode(int value) {
        val = value;
        left = null;
        right = null;
    } //treenode

    public List<List<Integer>> levelOrder(TreeNode root) {
        List<List<Integer>> result = new ArrayList<>();
        if (root == null) return result;

        Queue<TreeNode> queue = new LinkedList<>();
        queue.offer(root);

        while (!queue.isEmpty()) {
            int levelSize = queue.size();
            List<Integer> currentLevel = new ArrayList<>();

            for (int i = 0; i < levelSize; i++) {
                TreeNode node = queue.poll();
                currentLevel.add(node.val);
                if (node.left != null)
                    queue.offer(node.left);
                if (node.right != null)
                    queue.offer(node.right);
            } //for

            result.add(currentLevel);
        } //while

        return result;
    } //level order

    public static void main(String[] args) {
        TreeNode root = new TreeNode(4);
        root.left = new TreeNode(3);
        root.left.left = new TreeNode(1);
        root.right = new TreeNode(8);
        root.right.left = new TreeNode(5);
        root.right.right = new TreeNode(9);

        TreeNode tree = new TreeNode(3);
        List<List<Integer>> traversal = tree.levelOrder(root);
        System.out.println(traversal);
    }
}